### ISSUE TITLE
Clarifying whether join() might allocate

### ISSUE THREAD

**Issue by oconnor663** [rayon-rs/rayon#681]
The documentation for `scope` suggests that `join` makes exclusive use of the stack. I always assumed that meant "exclusive use of the stack, except for the first time you call it, when it will initialize the global thread pool." But looking through the code for `join`, it calls `crossbeam_deque::Worker::push`, which internally might call `resize` and allocate. Does that mean it's possible for `join` to allocate even after the first call, and to panic if that allocation fails? Would it be possible to make the allocation behavior explicit in the docs?

**Comment by cuviper** [comment]
Yes, it may allocate indirectly -- see also the discussion in #666.

I think the point of `scope`'s statement is that `Scope::spawn` *always* directly makes a heap allocation for the spawned task. This is in addition to whatever internal allocation might be required for the deques/queues. On the other hand, `join` never directly allocates anything, though it may still be affected by what crossbeam does.

I guess we can try to better clarify this in the docs, but I'm wary of getting too explicit about crossbeam's behavior, given that this isn't in our control.
